lib/task_manager.js

```diff
@@ -0,0 +1,67 @@
+class TaskManager {
+  constructor() {
+    this.map = {};
+    this.queue;
+    this.buffer = [];
+    this.lastPushedToQueue = 0;
+  }
+
+  async initQueue(maxConcurrentRequests) {
+    const PQueue = (await import('p-queue')).default;
+    this.queue = new PQueue({ concurrency: maxConcurrentRequests });
+    this.queue.on('completed', (data) => this.handleNewData(data));
+    // this.queue.on('error', err => { return Promise.reject(err.toString()); } );
+  }
+
+  clearQueue() {
+    this.taskManager.queue.pause();
+    this.taskManager.queue.clear();
+  }
+
+  #generateDefensiveCopy() {
+    const bufferCopy = [];
+    for (const data of this.buffer) bufferCopy.push(data);
+    return bufferCopy;
+  }
+
+  #shiftOriginalBuffer(shiftsCount) {
+    while (shiftsCount >= 0) {
+      this.buffer.shift();
+      shiftsCount--;
+    }
+  }
+
+  retrieveCompleted() {
+    const bufferCopy =  this.#generateDefensiveCopy();
+    let shiftsCount = bufferCopy.length - 1;
+    this.#shiftOriginalBuffer(shiftsCount);
+
+    return bufferCopy;
+  }
+
+  #pushAllEligable() {
+    while (this.map[this.lastPushedToQueue]) {
+      for (const data of this.map[this.lastPushedToQueue]) {
+        this.buffer.push(data);
+        this.lastProcessedPosition = {
+          primaryKey: this.buffer[this.buffer.length - 1].primaryKey,
+          blockNumber: this.buffer[this.buffer.length - 1].blockNumber
+        };
+      }
+
+      delete this.map[this.lastPushedToQueue];
+      this.lastPushedToQueue++;
+    }
+  }
+
+  handleNewData(newTransformedData) {
+    this.map[this.lastPushedToQueue] = newTransformedData;
```

Comment:
hmmm. Maybe I'm wrong but I think this doesn't work if we have our tasks completed out of order. i.e. [3,2,1] - this will update lastProcessedPosition to blockNumber `1`. right?